packages/@headlessui-react/src/components/menu/menu.tsx

```diff
@@ -719,66 +427,66 @@ function ItemsFn<TTag extends ElementType = typeof DEFAULT_ITEMS_TAG>(
 
       // @ts-expect-error Fallthrough is expected here
       case Keys.Space:
-        if (state.searchQuery !== '') {
+        if (machine.state.searchQuery !== '') {
           event.preventDefault()
           event.stopPropagation()
-          return dispatch({ type: ActionTypes.Search, value: event.key })
+          return machine.send({ type: ActionTypes.Search, value: event.key })
         }
       // When in type ahead mode, fallthrough
       case Keys.Enter:
         event.preventDefault()
         event.stopPropagation()
-        dispatch({ type: ActionTypes.CloseMenu })
-        if (state.activeItemIndex !== null) {
-          let { dataRef } = state.items[state.activeItemIndex]
+        machine.send({ type: ActionTypes.CloseMenu })
+        if (machine.state.activeItemIndex !== null) {
+          let { dataRef } = machine.state.items[machine.state.activeItemIndex]
           dataRef.current?.domRef.current?.click()
         }
-        restoreFocusIfNecessary(state.buttonElement)
+        restoreFocusIfNecessary(buttonElement)
```

Comment:
Shouldn't this be `machine.state.buttonElement`? It's closing over a potentially old value

Reply:
Yeah not a bad idea to use the `machine.state` in event listeners. Thanks!